vector-stores/spring-ai-azure-cosmos-db-store/src/main/java/org/springframework/ai/vectorstore/CosmosDBVectorStore.java

```diff
@@ -0,0 +1,318 @@
+		embedding.setDataType(CosmosVectorDataType.FLOAT32);
+		embedding.setDimensions(vectorDimensions);
+		embedding.setDistanceFunction(CosmosVectorDistanceFunction.COSINE);
+		embeddingPolicy.setCosmosVectorEmbeddings(Collections.singletonList(embedding));
+		collectionDefinition.setVectorEmbeddingPolicy(embeddingPolicy);
+
+		// set vector indexing policy
+		IndexingPolicy indexingPolicy = new IndexingPolicy();
+		indexingPolicy.setIndexingMode(IndexingMode.CONSISTENT);
+		ExcludedPath excludedPath = new ExcludedPath("/*");
+		indexingPolicy.setExcludedPaths(Collections.singletonList(excludedPath));
+		IncludedPath includedPath1 = new IncludedPath("/metadata/?");
+		IncludedPath includedPath2 = new IncludedPath("/content/?");
+		indexingPolicy.setIncludedPaths(ImmutableList.of(includedPath1, includedPath2));
+		CosmosVectorIndexSpec cosmosVectorIndexSpec = new CosmosVectorIndexSpec();
+		cosmosVectorIndexSpec.setPath("/embedding");
+		cosmosVectorIndexSpec.setType(CosmosVectorIndexType.DISK_ANN.toString());
+		indexingPolicy.setVectorIndexes(List.of(cosmosVectorIndexSpec));
+		collectionDefinition.setIndexingPolicy(indexingPolicy);
+
+		ThroughputProperties throughputProperties = ThroughputProperties.createManualThroughput(vectorStoreThoughput);
+		CosmosAsyncDatabase cosmosAsyncDatabase = cosmosClient.getDatabase(databaseName);
+		cosmosAsyncDatabase.createContainerIfNotExists(collectionDefinition, throughputProperties).block();
+		this.container = cosmosAsyncDatabase.getContainer(containerName);
+	}
+
+	@Override
+	public void close() {
+		if (cosmosClient != null) {
+			cosmosClient.close();
+			logger.info("Cosmos DB client closed successfully.");
+		}
+	}
+
+	private JsonNode mapCosmosDocument(Document document, float[] queryEmbedding) {
+		ObjectMapper objectMapper = new ObjectMapper();
+
+		String id = document.getId();
+		String content = document.getContent();
+
+		// Convert metadata and embedding directly to JsonNode
+		JsonNode metadataNode = objectMapper.valueToTree(document.getMetadata());
+		JsonNode embeddingNode = objectMapper.valueToTree(queryEmbedding);
+
+		// Create an ObjectNode specifically
+		ObjectNode objectNode = objectMapper.createObjectNode();
+
+		// Use put for simple values and set for JsonNode values
+		objectNode.put("id", id);
+		objectNode.put("content", content);
+		objectNode.set("metadata", metadataNode); // Use set to add JsonNode directly
+		objectNode.set("embedding", embeddingNode); // Use set to add JsonNode directly
+
+		return objectNode;
+	}
+
+	@Override
+	public void doAdd(List<Document> documents) {
+
+		// Create a list to hold both the CosmosItemOperation and the corresponding
+		// document ID
+		List<ImmutablePair<String, CosmosItemOperation>> itemOperationsWithIds = documents.stream().map(doc -> {
+			CosmosItemOperation operation = CosmosBulkOperations.getCreateItemOperation(
+					mapCosmosDocument(doc, this.embeddingModel.embed(doc.getContent())), new PartitionKey(doc.getId()));
```

Comment:
Have you explored the new `BatchingStrategy` API? https://docs.spring.io/spring-ai/reference/api/vectordbs.html#_batching_strategy. When you have many documents, the batching strategy allows you to embed batches of documents at once instead of going one by one. You need to extract the embedding call outside of where it is. Please take a look at the other vector stores for the usage pattern.

Reply:
Thanks @sobychacko. Is this a blocking change for the PR, in addition to rebasing? We will explore it.

Reply:
I have implemented it. Thank you.